### ISSUE TITLE
Add Stream.whenM for parity with ZIO.whenM

### ISSUE THREAD

**Issue by fanf** [zio/zio#4387]
I have the use case where I want to take a decision based on an effect. 
For ex: I want to walk a directory tree structure, and do things based on file existence. 

I find a way to do it like that:

```
    ZStream.fromEffect(
      Task.effect(file.exists).catchAll(ex =>
        logger.error(ex.message *> ZIO.succeed(false)
      )
    ).flatMap(cond =>
      if(!cond) ZStream.empty
      else {
        ZStream.fromIterator(Task.effect(file.children)).... etc
```

But it feels combersome. I would love to have a combinator looking like:

```
    ZStream.ifFromEffect(effect (: ZIO[Any, Nothing, Boolean]))(ifTrue: ZStream)(ifFalse: ZStream)
```

**Comment by jdegoes** [comment]
Perhaps what we need is `ZStream.whenM`. 

```scala
ZStream.whenM(Task(file.exists) orElseSucceed false)(ZStream.fromIterator(...))
```

**Comment by fanf** [comment]
I think `.whenM` would be a good addition. It's what I looked for first, for symmetry with `ZIO` case. And perhaps that the combinator is especially welcomed on that special case, and less so on the case where both true and false generate substream, because the quantity of boilerplate is almost negligible in that latter use case. 

So to sum-up: yes, `whenM`: take care of the case with more boilerplate and similar to `ZIO`.

**Comment by arnoldlacko** [comment]
I would like to take care of this one.

**Comment by adamgfraser** [comment]
Closed by #4542.
